sql/rpl_parallel.cc

```diff
@@ -3050,13 +3061,13 @@ rpl_parallel::stop_during_until()
 }
 
 
-bool
-rpl_parallel::workers_idle(Relay_log_info *rli)
+bool Relay_log_info::are_worker_threads_caught_up()
 {
-  mysql_mutex_assert_owner(&rli->data_lock);
-  return !rli->last_inuse_relaylog ||
-    rli->last_inuse_relaylog->queued_count ==
-    rli->last_inuse_relaylog->dequeued_count;
+  mysql_mutex_assert_owner(&data_lock);
+  if (!worker_threads_caught_up && (!last_inuse_relaylog ||
```

Comment:
Re. https://github.com/MariaDB/server/pull/4103#discussion_r2151239208, I still wonder if we can skip this null check at any of the two places.